index.js

```diff
@@ -9,42 +9,65 @@ var hasFlag = function (flag) {
 };
 
 module.exports = (function () {
-	if ('FORCE_COLOR' in process.env) {
-		return true;
+	var support = function (level) {
+		if (level <= 0) return null;
+		var result = {};
+		result.level = level;
+		result.has16 = result.hasBasic = level >= 1;
+		result.has256 = level >= 2;
+		result.has16m = level >= 3;
+		return result;
 	}
 
-	if (hasFlag('no-color') ||
-		hasFlag('no-colors') ||
-		hasFlag('color=false')) {
-		return false;
-	}
+	var result = (function () {
 
-	if (hasFlag('color') ||
-		hasFlag('colors') ||
-		hasFlag('color=true') ||
-		hasFlag('color=always')) {
-		return true;
-	}
+		if (hasFlag('color=16m') ||
+			hasFlag('color=full') ||
+			hasFlag('color=truecolor')) {
+			return support(3);
+		}
 
-	if (process.stdout && !process.stdout.isTTY) {
-		return false;
-	}
+		if (hasFlag('color=256')) {
+			return support(2);
+		}
 
-	if (process.platform === 'win32') {
-		return true;
-	}
+		if (hasFlag('color') ||
+			hasFlag('colors') ||
+			hasFlag('color=true') ||
+			hasFlag('color=always')) {
+			return support(1);
+		}
 
-	if ('COLORTERM' in process.env) {
-		return true;
-	}
+		if (process.stdout && !process.stdout.isTTY) {
+			return support(0);
+		}
 
-	if (process.env.TERM === 'dumb') {
-		return false;
-	}
+		if (process.platform === 'win32') {
+			return support(1);
+		}
+
+		if ('COLORTERM' in process.env) {
+			return support(1);
+		}
+
+		if (process.env.TERM === 'dumb') {
+			return support(0);
+		}
+
+		if (/^xterm-256(?:color)?/.test(process.env.TERM)) {
+			return support(2);
+		}
+
+		if (/^screen|^xterm|^vt100|color|ansi|cygwin|linux/i.test(process.env.TERM)) {
+			return support(1);
+		}
+
+		return support(0);
```

Comment:
Easier if all these return the number a level, and only use `support` in the last step.